Allow comments after object description

DavidFeldhoff opened the pull request:
Hi Johannes,

we have some objects which have a comment at the end of the first line.
if *any* file has a comment with a quote at the end of "object description", then the command "NAB: Find code source of current line" crashes.
![image](https://user-images.githubusercontent.com/53570297/143038920-2c88f4ff-5fe7-4407-8296-3de8b3ce9fcd.png)

In this PR I'm just changing the regex slightly, so that the 4th group is only found if a quote is **before** a `//` or `/*`

If that works, then I'm very glad to work with this command further :) Thanks for that!

Best regards
David

theschitz commented:
@DavidFeldhoff Seems like this should have been an issue, but I'll allow it for now :) In the mean time you could consider the failed test.

DavidFeldhoff commented:
I've fixed the test right away and added 4 new valid examples.
Locally the test ran successful now.
And I did not create an issue as I thought I could save you time with doing it by myself. But I can create an issue of course next time, if it suits you better. Totally fine for me :)

jwikman commented:
Hi @DavidFeldhoff
Nice seeing you here and thanks for contributing! :)

I'll have a look at this now. That part of the code is probably the oldest in this extension, so it are probably a lot of room for improvements. I believe that I learned quite a lot about typescript since then :)

But the workflow failed due to some issues that prettier found. You had escaped " in a '-surrounded string. I think you'll be fine if just remove the backslashes before " in those strings.

DavidFeldhoff commented:
Wow, you have an awesome pipeline built in checking for that! :)
Doesn't matter, if it works, it will help us pretty much. Just that command already is a BIG help :) And I have those pieces in my code for sure as well ^^
Thanks for taking your time so fast and I hope the code fits your requirements now. If not, just let me know and I'll improve it further.

jwikman commented:
Hehe, yes it's a few tests by now. But they really helps!  There are so many different variants of everything nowadays, so it's easier to write tests for those we can figure out then doing manual test...

We implemented prettier not that long ago, and I'm not used with it yet. It's quite picky. But so far we just obey and do what it says. 😅
You've got quite a few new warnings after your last change ;-)

theschitz commented:
> I've fixed the test right away and added 4 new valid examples. Locally the test ran successful now. And I did not create an issue as I thought I could save you time with doing it by myself. But I can create an issue of course next time, if it suits you better. Totally fine for me :)

It's always a "depends on" kinda thing with issues preceding PRs. If it needs discussion or a specification, it's probably an issue. If it's something small this I'd say a PR is ok. We don't have that many contributions in the form of pull requests, so yours is very, very welcome!  Glance though [CONTRIBUTING.md](https://github.com/jwikman/nab-al-tools/blob/master/CONTRIBUTING.md) if you haven't already. And don't worry, we are not strict, we appreciate the help, effort and ideas of others. So thank you :sunglasses:  :+1:

theschitz commented:
(off topic) @DavidFeldhoff I saw that you contributed to ps-xliff-sync. If your interested in running NAB AL Tools functions in a pipeline there's some features already implemented and an open issue on the matter #158.

DavidFeldhoff commented:
@jwikman You're right about the new warnings. And now I now what you mean with it is quite picky 😅
I installed Prettier now as well, so that I comply with it and your warnings stay at 0.
I hope that's it now 😀 
@theschitz Shame on me, I haven't read that. I've just created a Pull Request where it states that an issue would be nice, but is not mandatory (so it differs a bit to the lines in Contributing.md) which is why I continued with creating the PR. I'll be better next time :)

jwikman commented:
@DavidFeldhoff as I wrote, that part of code is old and the regex-expressions is far from optimal. They have been on my radar for refactoring for quite some time - maybe now is the time to do it.

I'll rip your test object descriptors to a new branch and see if I can clean up this regex mess a bit. I'll be back :)

DavidFeldhoff commented:
> (off topic) @DavidFeldhoff I saw that you contributed to ps-xliff-sync. If your interested in running NAB AL Tools functions in a pipeline there's some features already implemented and an open issue on the matter #158.

Yes, you're right. Currently we're using a self-build xlf-tool which is a bit like [AL Language Tools](https://marketplace.visualstudio.com/items?itemName=ClipDynamics.al-language-tools) to directly enter the translation like
```al
Caption = 'English Text', Comment = 'DEU="Deutscher Text"'
```

But AL Language Tool is not supported anymore and we do not want to maintain our own extension anymore as there are great open source ones. 
And then Rob (Xliff Sync) told me at the Directions that his extension supports it as well to generate the german xlf file based on the developer notes. That's the main reason why we're most probably switching to XliffSync as the way to do the translation worked quite well for us and our developers won't have to learn a totally new way again.

But one thing I'm missing at Xliff Sync: If I know that the translation is missing or needs adaptation then I want to jump to the source file directly. Honestly I don't know why.. Probalby just to check if it's not a false positive? :D
But that's the story why I created this PR as you're delivering that part. 

But if you would ask me what function I'd like to have in Pipelines which currently already exists in VS Code by your extension then I could think of the following: 
If there would be a powershell command (to use it in pipelines) which receives the "Xliff Generator" note e.g. `Table MyTable - Field MyField - Property Caption` and returns a link to the devops file like `https://dev.azure.com/[organization]/[project]/_git/[repositoryname]?path=/app/src/Attributes/src/codeunit/MyCodeunit.Codeunit.al&version=GBdevelop&line=11&lineEnd=11&lineStartColumn=9&lineEndColumn=14`, that would be really nice. But I guess that's too much to ask as the value added by this is not that big and the effort would probably be quite huge. So I'll leave the idea here, but please do not feel compelled in any way to implement this as it would really just be a nice-to-have :)

jwikman commented:
@DavidFeldhoff I created a new PR, #212.

I managed to clean up some code and have a much more robust matching for object names - without checking for " back and forth.

What I did not manage right now is to allow a block comment after the object descriptor. But honestly don't think that will be a problem in practice, don't you agree?

(We've banned /* comment */ in our company and I know many done the same)

I need to get some sleep now, will try to complete this tomorrow.

DavidFeldhoff commented:
Yes, we aren't using that as well. We just have normal comments ( // ) after the objectDescriptor. So if you're leaving it out, is totally fine for me :)
Thanks to both of you for the chat and the collaboration :)

jwikman commented:
@DavidFeldhoff If we had an issue, this conversation would have continued there... ;)

But anyways: I published a preview of this change (and other changes that are release pending). 

Download the vsix from https://github.com/jwikman/nab-al-tools/releases/tag/1.7.0-preview.1 and see if it works as you'd expect! Create an issue if you find any problems.

I'll close this PR now as the original issue should have been fixed by PR #212 .

Once again, thanks for contributing!
